### PR TITLE
Support comment formatting in experimental formatter

### LINKED ISSUE #1259: Provide comment formatting for experimental formatter
Adding on to https://github.com/eclipse/lemminx/pull/1254, there are basic aspects of comment formatting that need to be implemented for the experimental formatter. 

Related tests: https://github.com/JessicaJHee/lemminx/blob/a35d8d5cd5140e952b52306ed1332c0fc35ad75d/org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/format/experimental/XMLFormatterExperimentalTest.java#L506

### PR DESCRIPTION
**Pull request by JessicaJHee**
Fixes #1259 

Signed-off-by: Jessica He [jhe@redhat.com](mailto:jhe@redhat.com)

### CODE REVIEW

**Comment by JessicaJHee** [comment]
This PR current does not support max line width and the tests for it are disabled.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/DOMCommentFormatter.java [hunk 1]
```diff
@@ -46,46 +52,47 @@ public void formatComment(DOMComment commentNode, XMLFormattingConstraints paren
...
-						availableLineWidth--;
-						spaceStart = -1;
-						spaceEnd = -1;
-					}
+					spaceEnd = i;
+				}
+			} else {
+				// Ensure the edit is within the selected range
+				if (startRange != -1 && endRange != -1 && (startRange > spaceStart || endRange < spaceEnd)) {
+					return;
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/DOMCommentFormatter.java:73]: Have we have some tests with format with range for this line code?

Comment (JessicaJHee) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/DOMCommentFormatter.java:73]: I have added a test now :)

**Comment by datho7561** [comment]
One thing I noticed is that, if I have empty lines between comments, then `xml.format.preservedNewlines` isn't respected. For instance, if I set it to `2`, then format the following XML:

```xml
<root-element>
    <!-- aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa -->


    <!-- aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa -->


    <!-- aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa -->
</root-element>
```

Then it will remove the newlines in between the comments, wheras in the old formatter, it wouldn't

**Comment by angelozerr** [comment]
@JessicaJHee please fix conflicts.

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/format/experimental/XMLFormatterExperimentalTest.java [hunk 2]
```diff
@@ -628,51 +628,93 @@ public void testComment() throws BadLocationException {
...
+				"  comment comment comment comment comment comment comment comment --></a>";
+		SharedSettings settings = new SharedSettings();
+		assertFormat(content, expected, settings, //
+				te(0, 3, 1, 2, " "), //
+				te(1, 78, 1, 79, "\n  "), //
+				te(1, 150, 1, 151, "\n  "));
+		assertFormat(expected, expected, settings);
+	}
+
+	@Disabled
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/format/experimental/XMLFormatterExperimentalTest.java:697]: Whi it is disabled?

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/format/experimental/XMLFormatterJoinCommentLinesTest.java [hunk 3]
```diff
@@ -106,22 +107,24 @@ public void testCommentFormatSameLine() throws BadLocationException {
 
 	}
 
+	@Disabled
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/format/experimental/XMLFormatterJoinCommentLinesTest.java:110]: Whi it is disabled?

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/format/experimental/XMLFormatterExperimentalTest.java [hunk 4]
```diff
@@ -628,51 +628,93 @@ public void testComment() throws BadLocationException {
...
 				"2222" + lineSeparator() + //
 				"  3333 -->" + lineSeparator() + //
 				"</a>";
-		assertFormat(content, expected);
+		assertFormat(content, expected, //
+				te(2, 10, 2, 10, lineSeparator()));
+		assertFormat(expected, expected);
+	}
+
+	@Disabled
```
[the remark this answers is not in the document]
Comment (datho7561) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/format/experimental/XMLFormatterExperimentalTest.java:679]: This test is related to line length

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/DOMCommentFormatter.java [hunk 5]
```diff
@@ -44,55 +50,78 @@ public void formatComment(DOMComment commentNode, XMLFormattingConstraints paren
...
+				// Whitespaces
+				if (spaceStart == -1) {
+					spaceStart = i;
+				} else {
+					spaceEnd = i;
+				}
+			} else {
+				spaceEnd = i;
+				// Ensure the edit is within the selected range
+				if (startRange != -1 && endRange != -1 && (startRange > spaceStart || endRange < spaceEnd)) {
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/DOMCommentFormatter.java:86]: If I have:
```xml
<dress>


    <!--  as;ldkf j;lsa jdf;lsa jfd;lsa jfd;lakjs df;ljsfd;a jsdf
    ajs df
    asj dflk aslkjfas
    d kfaskdf jas jfdlkasj fd
    ka jdsfka
    sdf aksdfj aks
    jfdkajs fd
    kajs fdk  -->

<size></size>

</dress>
```

and `"xml.format.joinCommentLines": true,`, `"xml.format.maxLineWidth": 80,`

and I range format just the comment, it adjusts the indentation on `<size></size>`, even though that's not selected. I know this is a hard issue, but see if you can avoid adjusting the whitespace after the comment in this case.

**Review by datho7561 — CHANGES_REQUESTED** [review]
Noticed a few things, but it's looking very good so far!

**Review by datho7561 — CHANGES_REQUESTED** [review]

**Review by datho7561 — APPROVED** [review]
Looks good to me. Thank you Jessica!

**Pull request merged**